irohad/consensus/yac/impl/cluster_order.cpp

```diff
@@ -10,17 +10,35 @@ namespace iroha {
     namespace yac {
 
       boost::optional<ClusterOrdering> ClusterOrdering::create(
-          const std::vector<std::shared_ptr<shared_model::interface::Peer>>
-              &order) {
+          const std::vector<std::shared_ptr<shared_model::interface::Peer>> &order,
+          std::vector<size_t> const& peer_positions) {
+        if (order.empty()) {
+          return boost::none;
+        }
+        return ClusterOrdering(order, peer_positions);
+      }
+
+      boost::optional<ClusterOrdering> ClusterOrdering::create(
+          const std::vector<std::shared_ptr<shared_model::interface::Peer>> &order) {
         if (order.empty()) {
           return boost::none;
         }
         return ClusterOrdering(order);
       }
 
       ClusterOrdering::ClusterOrdering(
-          std::vector<std::shared_ptr<shared_model::interface::Peer>> order)
-          : order_(std::move(order)) {}
+          std::vector<std::shared_ptr<shared_model::interface::Peer>> const& order,
+          std::vector<size_t> const& peer_positions) {
+
+        order_.reserve(order.size());
+        for (auto const& i : peer_positions) {
+          order_.emplace_back(order[i]);
+        }
+      }
+
+      ClusterOrdering::ClusterOrdering(
+          std::vector<std::shared_ptr<shared_model::interface::Peer>> const& order)
+          : order_(order) { }
```

Comment:
From the point of view of algorithmic complexity, the move-constructor call have O(1). But in terms of performance, calling move-constructor is much more expensive than passing by reference. Resources will be spent to form stack frame, to destroy it and to execute a function call instruction, which can cause an additional delay because of L1i-cache miss while address jump. The transfer by reference mechanism does a great job of delivering data. Why do we need to spend resources without any profit of it?